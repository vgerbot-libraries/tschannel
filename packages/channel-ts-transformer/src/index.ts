import ts from 'typescript';

class ChannelProgramContext {
    public channelClassDeclaration?: ts.ImportSpecifier;
    public channelClassSymbol?: ts.Symbol;
    public channelInstanceDeclarations: Array<ts.Declaration> = [];
    public sourceFileNode!: ts.SourceFile;
}

export default function transformer(program: ts.Program): ts.TransformerFactory<ts.Node> {
    // console.log('create transformer', program);
    return (context: ts.TransformationContext) => {
        // console.log('transformer context', context);
        return (file: ts.Node) => {
            const programCtx = new ChannelProgramContext();
            programCtx.sourceFileNode = file as ts.SourceFile;
            return ts.visitEachChild(file, visitor, context);

            function visitor(node: ts.Node): ts.Node | Array<ts.Node> {
                const ret = visitNode(node, program, programCtx, context);
                if(!ret) {
                    return ts.visitEachChild(node, visitor, context);
                } else {
                    return ret;
                }
            }
        };
    };
}


function visitNode(
    node: ts.Node,
    program: ts.Program,
    programCtx: ChannelProgramContext,
    context: ts.TransformationContext
): undefined | ts.Node | Array<ts.Node> {
    const typeChecker = program.getTypeChecker();
    const factory = context.factory;
    if(ts.isPropertyAssignment(node)) {
        const initializer = node.initializer;
        if(ts.isNewExpression(initializer)) {
            const typeSymbol = typeChecker.getSymbolAtLocation(initializer.expression);
            if(typeSymbol === programCtx.channelClassSymbol) {
                const variableSymbol = typeChecker.getSymbolAtLocation(node.name);
                if(!variableSymbol) {
                    return;
                }
                programCtx.channelInstanceDeclarations.push(variableSymbol.valueDeclaration);
            }
        }
    } if((ts.isParameter(node) || ts.isPropertyDeclaration(node) || ts.isVariableDeclaration(node)) && node.type) {
        const type = node.type as ts.TypeReferenceNode;
        const typeSymbol = typeChecker.getSymbolAtLocation(type.typeName);
        if(typeSymbol === programCtx.channelClassSymbol) {
            const variableSymbol = typeChecker.getSymbolAtLocation(node.name);
            if(!variableSymbol) {
                return;
            }
            programCtx.channelInstanceDeclarations.push(variableSymbol.valueDeclaration);
        }
    } else if (ts.isImportDeclaration(node)) {
        if (!ts.isStringLiteral(node.moduleSpecifier)) {
            return;
        }
        const moduleName = node.moduleSpecifier.text;
        if (moduleName !== 'tschannel') {
            return;
        }
        const namedImports = node.importClause?.namedBindings;
        if (namedImports && ts.isNamedImports(namedImports)) {
            const channelClassDeclaration = namedImports.elements.find((it) => it.getText() === 'Channel');
            if (!channelClassDeclaration) {
                return;
            }
            programCtx.channelClassDeclaration = channelClassDeclaration;
            programCtx.channelClassSymbol = typeChecker.getSymbolAtLocation(
                channelClassDeclaration?.getChildAt(0)
            );
        }
    } else if (ts.isCallExpression(node)) {
        const propertyExpression = node.expression;
        if(ts.isPropertyAccessExpression(propertyExpression)) {
            const propertyName = propertyExpression.name.text;
            if(propertyName !== 'rclass') {
                return;
            }
            const ownerExpression = propertyExpression.expression;
            if(ts.isNewExpression(ownerExpression)) {
                const classSymbol = typeChecker.getSymbolAtLocation(ownerExpression.expression);
                if (classSymbol !== programCtx.channelClassSymbol) {
                    return;
                }
            } else {
                const symbol = typeChecker.getSymbolAtLocation(ownerExpression);
                if (!symbol) {
                    return;
                }
                const index = programCtx.channelInstanceDeclarations.indexOf(symbol.valueDeclaration);
                if (index < 0) {
                    return;
                }
            }
            const typeArgs = node.typeArguments;
            if (!typeArgs || typeArgs.length < 1) {
                return;
            }

            const typeNode = typeArgs[0];
            const classExpression = createRemoteClassExpression(typeNode, typeChecker, factory);
            let classIdArg = node.arguments[0];
            if(!classIdArg) {
                classIdArg = factory.createStringLiteral(typeNode.getText())
            }
            return factory.createCallExpression(node.expression, [], [
                classIdArg,
                classExpression
            ]);
        }
    }
}
function createRemoteClassExpression(typeNode: ts.TypeNode, typeChecker: ts.TypeChecker, factory: ts.NodeFactory) {
    const className = ts.factory.createUniqueName(typeNode.getText());
    const type = typeChecker.getTypeFromTypeNode(typeNode);
    const members = typeChecker.getPropertiesOfType(type);
    const classMembers = members
        .filter(it => ts.isMethodSignature(it.valueDeclaration))
        .map(it => {
            return factory.createMethodDeclaration(
                [],
                [],
                undefined,
                it.getName(),
                undefined,
                [],
                [],
                undefined,
                factory.createBlock([], false)
            );
        });
    const classExpression = factory.createClassExpression(
        [],
        [],
        className,
        [],
        [ts.factory.createHeritageClause(ts.SyntaxKind.ImplementsKeyword, [])],
        classMembers
    );
    return classExpression;
}

