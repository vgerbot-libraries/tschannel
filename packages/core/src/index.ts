export { rmethod } from './annotations/rmethod';
// export { get_class } from './annotations/get_class';
export { Channel, Remote } from './Channel';
export { ParameterType } from './types/ParameterType';
export { RMIMethodMetadata } from './metadata/RMIMethodMetadata';
export {
    ParallelDataDistributor,
    ParallelDataCombiner,
    ParallelCommunicator
} from './communicators/ParallelCommunicator';

export { WebWorkerCommunicator } from './communicators/WebWorkerCommunicator';
export { WebWorkerScopeCommunicator } from './communicators/WebWorkerScopeCommunicator';
export { WindowChannelCommunicator } from './communicators/WindowChannelCommunicator';
export { InvokeMethodData } from './types/InvokeMethodData';
export { Returning } from './types/Returning';
export { default as InvokeMethodPayload } from './foundation/InvokeMethodPayload';
export { default as MethodReturningPayload } from './foundation/MethodReturningPayload';
export { StorageChannelCommunicator } from './communicators/StorageChannelCommunicator';
export { Adaptor } from './Adaptor';
export { Destructible } from './foundation/Destructible';

export * from './functional';
