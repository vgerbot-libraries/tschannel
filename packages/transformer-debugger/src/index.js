"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tschannel_1 = require("tschannel");
const channelId = 'test-transformer';
const communicator1 = new tschannel_1.StorageChannelCommunicator(localStorage, channelId);
const channel = new tschannel_1.Channel(channelId, communicator1);
new tschannel_1.Channel(channelId, communicator1).rclass();
// const RRemoteAPI = channel.rclass<RemoteAPI>();
channel.rclass();
