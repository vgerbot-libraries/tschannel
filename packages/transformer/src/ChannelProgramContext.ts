import ts from 'typescript';

export class ChannelProgramContext {
    constructor(private readonly typeChecker: ts.TypeChecker) {}
    public channelMethodSymbol?: ts.Symbol;
    public channelClassSymbol?: ts.Symbol;
    public variablesMap = new Map<ts.Type, ts.VariableDeclaration>();
    public channel_variables = new Set<ts.Symbol>();

    is_accessing_get_class_method(callExpression: ts.CallExpression, propertyExpression: ts.PropertyAccessExpression) {
        const propertyName = propertyExpression.name.text;
        if (propertyName !== 'get_class') {
            return false;
        }
        if(callExpression.typeArguments?.length !== 1) {
            return false;
        }
        const expressionType = this.typeChecker.getTypeAtLocation(propertyExpression.expression);
        if(expressionType && expressionType.getSymbol() === this.channelClassSymbol) {
            return true;
        }
        const LeftHandSideExpressionSymbol = this.typeChecker.getSymbolAtLocation(propertyExpression.expression);
        return !!LeftHandSideExpressionSymbol && this.channel_variables.has(LeftHandSideExpressionSymbol);
    }

    recordChannelVariableByBinaryExpression(node: ts.BinaryExpression) {
        const typeChecker = this.typeChecker;
        const variables = this.channel_variables;

        const variableSymbol = typeChecker.getSymbolAtLocation(node.left);
        if(!variableSymbol) {
            return;
        }
        if(this.isChannelInstanceInitializerExpression(node.right)) {
            variables.add(variableSymbol);
        }
    }
    recordChannelVariableIfPossible(node: ts.VariableDeclaration) {
        const typeChecker = this.typeChecker;
        const variables = this.channel_variables;
        const initializer = node.initializer;
        if(!initializer) {
            return;
        }
        const variableSymbol = typeChecker.getSymbolAtLocation(node.name);
        if(!variableSymbol) {
            return;
        }
        if(this.isChannelInstanceInitializerExpression(initializer)) {
            variables.add(variableSymbol);
        }
    }
    private isChannelInstanceInitializerExpression(initializer: ts.Expression) {
        const typeChecker = this.typeChecker;
        if(ts.isNewExpression(initializer)) {
            const classSymbol = typeChecker.getSymbolAtLocation(initializer.expression);
            if(classSymbol === this.channelClassSymbol) {
                return true;
            }
        } else if(ts.isCallExpression(initializer)) {
            let expression = initializer.expression;
            while(true) {
                if(ts.isCallExpression(expression)) {
                    expression = expression.expression;
                } else if(ts.isPropertyAccessExpression(expression)) {
                    expression = expression.expression;
                } else {
                    break;
                }
            }
            if(ts.isIdentifier(expression)) {
                const symbol = typeChecker.getSymbolAtLocation(expression);
                if(!!this.channelMethodSymbol && this.channelMethodSymbol === symbol) {
                    return true;
                }
                if(symbol) {
                    return this.channel_variables.has(symbol);
                }
            }
        }
        return false;
    }
    recordChannelSymbolIfPossible(node: ts.ImportDeclaration) {
        const namedBindings = node.importClause?.namedBindings;
        if(namedBindings && ts.isNamedImports(namedBindings)) {
            const importElementsArray = namedBindings.elements.map(it => {
                const name = it.propertyName ? it.propertyName.text : it.name.text;
                return { name, importSpecifier: it, symbol: this.typeChecker.getSymbolAtLocation(it.name) };
            })
            const { symbol: channelMethodSymbol } = importElementsArray.find(it => it.name === 'channel') || {};
            if(channelMethodSymbol) {
                this.channelMethodSymbol = channelMethodSymbol;
            }
            const { symbol: channelClassSymbol } = importElementsArray.find(it => it.name === 'Channel') || {};
            if(channelClassSymbol) {
                this.channelClassSymbol = channelClassSymbol;
            }
        }
    }
}
