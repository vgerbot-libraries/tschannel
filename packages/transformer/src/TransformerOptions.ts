// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface TransformerOptions {
    classIdStrategy: 'self' | 'first-interface' | 'first-parent';
}
