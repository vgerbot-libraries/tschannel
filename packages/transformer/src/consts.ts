export const CHANNEL_MODULE_NAME = '@vgerbot/channel';
export const DEFAULT_TRANSFORMER_OPTIONS = {
    // EMPTY
};
