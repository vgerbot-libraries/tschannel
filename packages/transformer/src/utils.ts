import ts from 'typescript';
import { CHANNEL_MODULE_NAME } from './consts';

export function getTypeNodeDecration(typeNodeObj: ts.Type) {
    let declarations;
    if (!!typeNodeObj.aliasSymbol) {
        declarations = typeNodeObj.aliasSymbol.getDeclarations();
    } else {
        declarations = typeNodeObj.getSymbol()?.getDeclarations();
    }

    return declarations ? declarations[0] : undefined;
}

export function createMemberNamesvariable(variableName: string, memberNames: string[], factory: ts.NodeFactory) {
    const memberNameLiterals = memberNames.map((it) => {
        return factory.createStringLiteral(it);
    });

    const membersArrayExpression = factory.createArrayLiteralExpression(memberNameLiterals);

    const membersVariableName = factory.createUniqueName(variableName);

    return factory.createVariableDeclaration(
        membersVariableName,
        undefined,
        undefined,
        membersArrayExpression
    );
}

export function getMethodMembersFrom(typeChecker: ts.TypeChecker, typeNode: ts.Type) {
    const members = typeChecker.getPropertiesOfType(typeNode);
    return members.filter((it) => it.valueDeclaration !== undefined && ts.isMethodSignature(it.valueDeclaration));
}

export function getSymbolFromChannel(program: ts.Program, name: string) {
    const typeChecker = program.getTypeChecker();

    const source = program.getSourceFile(`/${CHANNEL_MODULE_NAME}.ts`);
    if (!source) {
        return;
    }
    const symbol = typeChecker.getSymbolAtLocation(source);
    if (!symbol) {
        return;
    }
    const exports = typeChecker.getExportsOfModule(symbol);
    const ChannelClassSymbol = exports.find((it) => {
        return it.getName() === 'Channel';
    });
    return ChannelClassSymbol?.members?.get(name as ts.__String);
}
