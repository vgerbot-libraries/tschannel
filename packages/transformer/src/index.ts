import { CHANNEL_MODULE_NAME, DEFAULT_TRANSFORMER_OPTIONS } from './consts';
import { createMemberNamesvariable, getMethodMembersFrom, getTypeNodeDecration } from './utils';
import ts, { factory } from 'typescript';
import { ChannelProgramContext } from './ChannelProgramContext';
import { TransformerOptions } from './TransformerOptions';

export { TransformerOptions };

export function channelTransformerFactory(
    program: ts.Program,
    options?: Partial<TransformerOptions>
): ts.TransformerFactory<ts.SourceFile> {
    const resolvedOptions = {
        ...DEFAULT_TRANSFORMER_OPTIONS,
        ...(options || {}),
    };
    return (context: ts.TransformationContext) => {
        return (file: ts.Node) => {
            const programCtx = new ChannelProgramContext(program.getTypeChecker());

            const sourceFileNode = ts.visitEachChild(file, visitor, context) as ts.SourceFile;

            const variableDeclarations = Array.from(programCtx.variablesMap.values());
            const variableStatements =
                variableDeclarations.length > 0 ? [factory.createVariableStatement([], variableDeclarations)] : [];
            return ts.factory.updateSourceFile(sourceFileNode, [...variableStatements, ...sourceFileNode.statements]);

            function visitor(node: ts.Node): ts.Node | Array<ts.Node> {
                const ret = visitNode(node, program, programCtx, context, resolvedOptions);
                if (!ret) {
                    return ts.visitEachChild(node, visitor, context);
                } else {
                    return ret;
                }
            }
        };
    };
}

function visitNode(
    node: ts.Node,
    program: ts.Program,
    programCtx: ChannelProgramContext,
    context: ts.TransformationContext,
    options: Partial<TransformerOptions>
): undefined | ts.Node | Array<ts.Node> {
    const variablesMap = programCtx.variablesMap;
    const typeChecker = program.getTypeChecker();
    const factory = context.factory;
    if (ts.isImportDeclaration(node) && ts.isStringLiteralLike(node.moduleSpecifier)) {
        const moduleName = node.moduleSpecifier.text;
        if (moduleName !== CHANNEL_MODULE_NAME) {
            return;
        }
        programCtx.recordChannelSymbolIfPossible(node);
    } else if (ts.isVariableDeclaration(node)) {
        programCtx.recordChannelVariableIfPossible(node);
    } else if (ts.isBinaryExpression(node)) {
        programCtx.recordChannelVariableByBinaryExpression(node);
    } else if (ts.isCallExpression(node)) {
        const propertyExpression = node.expression;
        if (ts.isPropertyAccessExpression(propertyExpression)) {
            if (programCtx.isAccessingTheGetClassMethod(node, propertyExpression)) {
                return handleGetClassMethod(node, typeChecker, factory, variablesMap);
            } else if (programCtx.isAccessingDefClassMethod(node, propertyExpression)) {
                return handleDefClassMethod(node, typeChecker, options, factory);
            } else if (programCtx.isAccessingDefMethodMethod(node, propertyExpression)) {
                return handleDefMethodMethod(node, factory);
            }
        }
    }
}

function handleGetClassMethod(
    node: ts.CallExpression,
    typeChecker: ts.TypeChecker,
    factory: ts.NodeFactory,
    variablesMap: Map<ts.Type, ts.VariableDeclaration>
) {
    const typeArgs = node.typeArguments;
    if (!typeArgs || typeArgs.length < 1) {
        return;
    }

    const typeNode = typeArgs[0];
    const typeNodeObj = typeChecker.getTypeFromTypeNode(typeNode);

    const typeNodeDeclaration = getTypeNodeDecration(typeNodeObj);
    if (!typeNodeDeclaration) {
        return;
    }
    if (node.arguments.length > 1) {
        return;
    }

    const classIdArg = node.arguments[0] || factory.createStringLiteral(typeNode.getText());

    let interfaceNode: ts.Type;
    let memberNames: undefined | string[];
    if (ts.isClassDeclaration(typeNodeDeclaration)) {
        const modifiers = (typeNodeDeclaration as ts.ClassDeclaration).modifiers;
        const isAbstract = !!modifiers && modifiers.some((it) => it.kind === ts.SyntaxKind.AbstractKeyword);
        if (!isAbstract) {
            return factory.createCallExpression(
                node.expression,
                [],
                [classIdArg, factory.createRegularExpressionLiteral(typeNode.getText())]
            );
        }
        interfaceNode = typeChecker.getTypeAtLocation(typeNodeDeclaration);
        memberNames = typeNodeDeclaration.members
            .filter((it) => !!it.name)
            .map((it) => it.name?.getText())
            .filter(Boolean) as string[];
    } else {
        interfaceNode = typeChecker.getTypeFromTypeNode(typeNode);
    }
    if (!interfaceNode) {
        return;
    }
    let variable = variablesMap.get(interfaceNode);
    if (!variable) {
        if (!memberNames || memberNames.length === 0) {
            memberNames = getMethodMembersFrom(typeChecker, interfaceNode).map((it) => it.getName());
        }
        variable = variable || createMemberNamesvariable(typeNode.getText() + 'Members', memberNames, factory);
        variablesMap.set(interfaceNode, variable);
    }

    return factory.createCallExpression(node.expression, [], [classIdArg, variable.name as ts.Identifier]);
}


function handleDefClassMethod(
    node: ts.CallExpression,
    typeChecker: ts.TypeChecker,
    options: Partial<TransformerOptions>,
    factory: ts.NodeFactory
) {
    const classIdentifier = node.arguments[0] as ts.Identifier;
    const classSymbol = typeChecker.getSymbolAtLocation(classIdentifier);
    if (!classSymbol) {
        return;
    }
    let classId!: string;
    if (options.classIdStrategy === 'first-interface') {
        const classDec = classSymbol.valueDeclaration as ts.ClassDeclaration;
        if (classDec.heritageClauses && classDec.heritageClauses.length > 0) {
            for (const clause of classDec.heritageClauses) {
                const expression = clause.types[0]?.expression;
                const symbol = expression && typeChecker.getSymbolAtLocation(expression);
                const declarations = symbol?.getDeclarations();
                const declaration = declarations ? declarations[0] : undefined;
                if (!!symbol && !!declaration && ts.isInterfaceDeclaration(declaration)) {
                    classId = symbol.getName();
                    break;
                }
            }
        }
    } else if (options.classIdStrategy === 'first-parent') {
        const classDec = classSymbol.valueDeclaration as ts.ClassDeclaration;
        if (classDec.heritageClauses && classDec.heritageClauses.length > 0) {
            const firstHeritageClause = classDec.heritageClauses[0];
            const expression = firstHeritageClause?.types[0].expression;
            const symbol = expression && typeChecker.getSymbolAtLocation(expression);
            const declarations = symbol?.getDeclarations();
            const declaration = declarations ? declarations[0] : undefined;
            if (!!symbol && !!declaration && ts.isInterfaceDeclaration(declaration)) {
                classId = symbol.getName();
            }
        }
    }

    if (!classId) {
        classId = classSymbol.getName();
    }
    const classIdArg = factory.createStringLiteral(classId);
    return factory.createCallExpression(node.expression, [], [classIdArg, classIdentifier]);
}

function handleDefMethodMethod(node: ts.CallExpression, factory: any) {
    const arg0 = node.arguments[0];
    if (ts.isIdentifier(arg0)) {
        const methodId = factory.createStringLiteral(arg0.text);
        return factory.createCallExpression(node.expression, [], [methodId, arg0]);
    } else if (ts.isFunctionExpression(arg0)) {
        const methodId = factory.createStringLiteral(arg0.name!.text);
        return factory.createCallExpression(node.expression, [], [methodId, arg0]);
    }
}
