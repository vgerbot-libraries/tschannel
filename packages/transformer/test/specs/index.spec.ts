import { loadFixtures, loadSpecialFixtures } from '../common/load-fixture';
import { TransformerOptions, channelTransformerFactory } from '../../src';

import { transpile } from '../common/compiler';

interface TestCase {
    source: string;
    filepath: string;
    name: string;
    only?: boolean;
    skip?: boolean;
    shouldThrow?: boolean | string;
    transformerOptions?: Partial<TransformerOptions>;
}

interface SpecialFixture {
    file: string;
    name: string;
    only?: boolean;
    skip?: boolean;
    shouldThrow?: boolean | string;
    transformerOptions?: Partial<TransformerOptions>;
}

describe('@vgerbot/channel-transformer', () => {
    const specialFixtures: SpecialFixture[] = [
        {
            file: 'normal',
            name: 'should run normally with the transformer',
        },
        {
            file: 'alias',
            name: 'should be able to detect the module alias',
        },
        {
            file: 'exclude 2 arguments',
            name: 'should be able to exclude methods calling that has more than 2 arguments',
        },
        {
            file: 'exclude without type arguments',
            name: 'should be able to exclude methods without type arguments',
        },
        {
            file: 'reference class',
            name: 'should be able to reference the class',
        },
        {
            file: 'reuse converted variables',
            name: 'should reuse converted variables',
        },
        {
            file: 'class id strategy first-interface',
            name: 'should use the name of the first parent interface as the class id',
            transformerOptions: {
                classIdStrategy: 'first-interface',
            },
        },
        {
            file: 'class id strategy first-parent',
            name: 'should use the name of the first parent class as the class id',
            transformerOptions: {
                classIdStrategy: 'first-parent',
            },
        },
        {
            file: 'detect anonymous arrow functions',
            name: 'should throw when passing a anonymous arrow function to Channel.def_method',
            shouldThrow: 'The first function parameter of Channel.def_method cannot be anonymous',
        },
        {
            file: 'detect anonymous functions',
            name: 'should throw when passing a anonymous function to Channel.def_method',
            shouldThrow: 'The first function parameter of Channel.def_method cannot be anonymous',
        },
    ];

    specialFixtures
        .map((it) => {
            return {
                ...it,
                source: loadSpecialFixtures(it.file).source,
                filepath: `fixtures/special/${it.file}.ts`,
            } as TestCase;
        })
        .concat(
            loadFixtures().map((it) => {
                return {
                    ...it,
                    name: `should transform "${it.filepath}" correctly`,
                    only: false,
                    skip: false,
                };
            })
        )
        .forEach(({ filepath, source, name, only, skip, transformerOptions, shouldThrow }) => {
            let callback: jest.ProvidesCallback = () => {
                const output = transpile(filepath, source, {
                    channelTransformer: channelTransformerFactory,
                    channelTransformerOptions: transformerOptions,
                });
                expect(output).toMatchSnapshot();
            };
            if (shouldThrow) {
                const test_case = callback;
                if (typeof shouldThrow === 'string') {
                    callback = () => {
                        expect(test_case).toThrowError(shouldThrow);
                    };
                } else {
                    callback = () => {
                        expect(test_case).toThrow();
                    };
                }
            }
            if (only) {
                it.only(name, callback);
            } else if (skip) {
                it.skip(name, callback);
            } else {
                it(name, callback);
            }
        });
});
