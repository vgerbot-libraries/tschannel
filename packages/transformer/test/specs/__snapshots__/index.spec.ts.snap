// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@vgerbot/channel-transformer should be able to detect the module alias 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var RemoteAPIMembers_1 = [\\"method\\"];
const channel_1 = require(\\"@vgerbot/channel\\");
const channel = new channel_1.Channel();
channel.get_class(\\"RemoteAPI\\", RemoteAPIMembers_1);
{ }
"
`;

exports[`@vgerbot/channel-transformer should be able to exclude methods calling that has more than 2 arguments 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var ;
const channel_1 = require(\\"@vgerbot/channel\\");
const channel = new channel_1.Channel();
channel.get_class('RemoteAPI1', class {
    method() { }
});
{ }
"
`;

exports[`@vgerbot/channel-transformer should be able to exclude methods without type arguments 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var ;
const channel_1 = require(\\"@vgerbot/channel\\");
const channel = new channel_1.Channel();
channel.get_class('RemoteAPI', class RemoteAPI {
    method() { }
});
"
`;

exports[`@vgerbot/channel-transformer should be able to reference the class 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var ;
const channel_1 = require(\\"@vgerbot/channel\\");
const channel = new channel_1.Channel();
channel.get_class(\\"RemoteAPI\\", RemoteAPI);
class RemoteAPI {
    method() { }
}
"
`;

exports[`@vgerbot/channel-transformer should be able to transform the absrtact class 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var RemoteAPIMembers_1 = [\\"method\\", \\"method2\\"];
const channel_1 = require(\\"@vgerbot/channel\\");
const channel = new channel_1.Channel();
channel.get_class(\\"RemoteAPI\\", RemoteAPIMembers_1);
class RemoteAPI {
    method() { }
}
"
`;

exports[`@vgerbot/channel-transformer should correctly reuse converted member name array variables 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var RemoteAPIMembers_1 = [\\"method\\", \\"method2\\"], RemoteAPIMembers_2 = [\\"method\\"], InnerRemoteAPIMembers_1 = [\\"method\\"];
const channel_1 = require(\\"@vgerbot/channel\\");
const channel = new channel_1.Channel();
channel.get_class(\\"RemoteAPI\\", RemoteAPIMembers_1);
channel.get_class(\\"RemoteAPI\\", RemoteAPIMembers_1);
class RemoteAPI {
    method() { }
}
function scope_a() {
    channel.get_class(\\"RemoteAPI\\", RemoteAPIMembers_2);
    channel.get_class(\\"RemoteAPI\\", RemoteAPIMembers_2);
    function scope_inner_a() {
        channel.get_class(\\"InnerRemoteAPI\\", InnerRemoteAPIMembers_1);
        channel.get_class(\\"RemoteAPI\\", RemoteAPIMembers_2);
    }
}
"
`;

exports[`@vgerbot/channel-transformer should run normally with the transformer 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.x = void 0;
var ;
const channel_1 = require(\\"@vgerbot/channel\\");
exports.x = 'string';
console.info(channel_1.Channel);
"
`;

exports[`@vgerbot/channel-transformer should transform the interface parameter correctly 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var RemoteAPIMembers_1 = [\\"method\\"];
const channel_1 = require(\\"@vgerbot/channel\\");
const channel = new channel_1.Channel();
channel.get_class(\\"RemoteAPI\\", RemoteAPIMembers_1);
{ }
"
`;
